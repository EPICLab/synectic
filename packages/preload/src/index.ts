/**
 * @module preload
 */

export {fileOpen, fileSave} from './dialogs';
export {flattenArray, flattenObject} from './flatten';
export {
  add,
  catFile,
  checkGitVersion,
  checkRefFormat,
  checkUnmergedBranch,
  checkUnmergedPath,
  checkoutBranch,
  checkoutDetached,
  checkoutNewBranch,
  checkoutPathspec,
  cloneRepo,
  commit,
  copyBranch,
  createBranch,
  deleteBranch,
  errorLogger,
  extractFromURL,
  extractRepoName,
  fetchMergingBranches,
  fileStatus,
  filterIgnore,
  getBranchRoot,
  getConfig,
  getCredentials,
  getIgnore,
  getMatchUp,
  getRemote,
  getRemoteConfig,
  getRoot,
  getWorktreePaths,
  isValidRepository,
  isValidRepositoryURL,
  listBranch,
  log,
  mergeBase,
  mergeBranch,
  mergeInProgress,
  moveBranch,
  processBranchOutput,
  processCatFileOutput,
  processLogOutput,
  processMergeBaseOutput,
  processMergeOutput,
  processRemoteOutput,
  processShowBranchOutput,
  processStatusCode,
  processWorktreeOutput,
  reset,
  resolveURL,
  restore,
  revList,
  revParse,
  revParseArgs,
  setConfig,
  setUpstreamBranch,
  showBranch,
  unsetUpstreamBranch,
  worktreeAdd,
  worktreeList,
  worktreeLock,
  worktreeMove,
  worktreePrune,
  worktreeRemove,
  worktreeRepair,
  worktreeStatus,
  worktreeUnlock,
  type Conflict,
  type GitConfig,
  type MergeAction,
  type WorktreePaths,
} from './git';
export {
  decompressBinaryObject,
  execute,
  extractDirname,
  extractExtension,
  extractFilename,
  extractStats,
  filterReadArray,
  getDescendantPaths,
  isDescendant,
  isDirectory,
  isEqualPaths,
  readDirAsync,
  readDirAsyncDepth,
  readFileAsync,
  replaceExt,
  validateFileName,
  writeFileAsync,
} from './io';
export {
  clipboard,
  cwd,
  dirname,
  join,
  normalize,
  openExternal,
  platform,
  relative,
  sep,
  sha256sum,
  uuid,
} from './platform';
export {
  asyncFilter,
  delay,
  deserialize,
  equalArrayBuffers,
  equalArrays,
  equalMaps,
  filterObject,
  filteredArrayEquality,
  filteredObjectEquality,
  getConflictingChunks,
  getRandomInt,
  hasUpdates,
  isDefined,
  isFilled,
  isModified,
  isNumber,
  isPresent,
  isStaged,
  isUnmerged,
  isWrappedQuote,
  objectEntries,
  objectifyPath,
  partition,
  removeDuplicates,
  removeNullableProperties,
  removeObjectProperty,
  removeUndefined,
  symmetrical,
  toArrayBuffer,
} from './utils';
export {versions} from './versions';
