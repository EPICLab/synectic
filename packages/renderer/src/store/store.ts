import {combineReducers, configureStore} from '@reduxjs/toolkit';
import branchesReducer from './slices/branches';
import cardsReducer from './slices/cards';
import commitsReducer from './slices/commits';
import filetypesReducer from './slices/filetypes';
import metafilesReducer from './slices/metafiles';
import modalsReducer from './slices/modals';
import reposReducer from './slices/repos';
import stacksReducer from './slices/stacks';
import {listenerMiddleware} from './listenerMiddleware';

export const rootReducer = combineReducers({
  branches: branchesReducer,
  cards: cardsReducer,
  commits: commitsReducer,
  filetypes: filetypesReducer,
  metafiles: metafilesReducer,
  modals: modalsReducer,
  repos: reposReducer,
  stacks: stacksReducer,
});

const store = configureStore({
  reducer: rootReducer,
  devTools: true,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware().concat([listenerMiddleware.middleware]),
});

export type RootState = ReturnType<typeof rootReducer>;
export type AppDispatch = typeof store.dispatch;

export default {store: store};
