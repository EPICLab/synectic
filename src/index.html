<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>Synectic Development Environment</title>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>
