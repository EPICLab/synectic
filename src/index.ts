import { app, BrowserWindow, dialog, ipcMain } from 'electron';
import installExtension, { REDUX_DEVTOOLS, REACT_DEVELOPER_TOOLS } from 'electron-devtools-installer';
// This allows TypeScript to pick up the magic constant that's auto-generated by Forge's Webpack
// plugin that tells the Electron app where to look for the Webpack-bundled app code (depending on
// whether you're running in development or production).
declare const MAIN_WINDOW_WEBPACK_ENTRY: string;

// Handle creating/removing shortcuts on Windows when installing/uninstalling.
if (require('electron-squirrel-startup')) { // eslint-disable-line global-require
  app.quit();
}

const createWindow = (): void => {
  // Create the browser window.
  const mainWindow = new BrowserWindow({
    height: 1200,
    width: 1400,
    webPreferences: {
      contextIsolation: false,
      nodeIntegration: true,
      webviewTag: true
    }
  });

  // and load the index.html of the app.
  mainWindow.loadURL(MAIN_WINDOW_WEBPACK_ENTRY);

  // Open the DevTools.
  if (process.env.NODE_ENV !== 'production') {
    mainWindow.webContents.openDevTools();
  }
};

// This method will be called when Electron has finished
// initialization and is ready to create browser windows.
// Some APIs can only be used after this event occurs.
app.on('ready', createWindow);
app.whenReady().then(() => {
  installExtension([REDUX_DEVTOOLS, REACT_DEVELOPER_TOOLS])
    .then((name) => console.log(`Added extension: ${name}`))
    .catch((err) => console.log('An error occurred: ', err))
});

// Handle renderer requests for Electron `showOpenDialog` and returns filepath results.
ipcMain.handle('fileOpenDialog', async (_event, properties) => {
  return await dialog.showOpenDialog({ properties: [...properties, 'multiSelections'] });
});

// Handle renderer requests for Electron `showSaveDialog` and returns the results value.
ipcMain.handle('fileSaveDialog', async (_event, { defaultPath, buttonLabel, properties }) => {
  return await dialog.showSaveDialog({ defaultPath, buttonLabel, properties });
});

// Quit when all windows are closed, except on macOS. There, it's common
// for applications and their menu bar to stay active until the user quits
// explicitly with Cmd + Q.
app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

app.on('activate', () => {
  // On OS X it's common to re-create a window in the app when the
  // dock icon is clicked and there are no other windows open.
  if (BrowserWindow.getAllWindows().length === 0) {
    createWindow();
  }
});

// In this file you can include the rest of your app's specific main process
// code. You can also put them in separate files and import them here.
