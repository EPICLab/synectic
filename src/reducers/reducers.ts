import { combineReducers } from 'redux';
import {
  // ADD_TODO,
  // TOGGLE_TODO,
  SET_VISIBILITY_FILTER,
  VisibilityFilters,
  ActionType
} from '../actions/actions';

const { SHOW_ALL } = VisibilityFilters;

// type TodoItem = {
//   text: string;
//   completed: boolean;
// }

// type TodoState = {
//   visibilityFilter: string;
//   todos: TodoItem[];
// }

function visibilityFilter(state = SHOW_ALL, action: ActionType) {
  switch (action.type) {
    case SET_VISIBILITY_FILTER:
      return action.filter;
    default:
      return state;
  }
}

// function todos(state: TodoItem[] = [], action: ActionType) {
//   switch (action.type) {
//     case ADD_TODO:
//       return [
//         ...state,
//         {
//           text: action.text,
//           completed: false
//         }
//       ];
//     case TOGGLE_TODO:
//       return state.map((todo, index) => {
//         if (index === action.index) {
//           return Object.assign({}, todo, {
//             completed: !todo.completed
//           });
//         }
//         return todo;
//       });
//     default:
//       return state;
//   }
// }

const todoApp = combineReducers({
  visibilityFilter
});

export default todoApp;