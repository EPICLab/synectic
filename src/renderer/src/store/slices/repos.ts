import { createEntityAdapter, createSlice } from '@reduxjs/toolkit';
import { Repository } from 'types/repo';

export const repoAdapter = createEntityAdapter<Repository>();
repoAdapter.addOne;

export const repoSlice = createSlice({
  name: 'repos',
  initialState: repoAdapter.getInitialState(),
  reducers: {
    repoAdded: repoAdapter.addOne,
    repoRemoved: repoAdapter.removeOne,
    repoUpdated: repoAdapter.upsertOne
  }
});

export const { repoAdded, repoRemoved, repoUpdated } = repoSlice.actions;

export default repoSlice.reducer;
