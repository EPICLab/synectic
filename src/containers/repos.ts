import * as fs from 'fs-extra';
import * as isogit from 'isomorphic-git';
import { PathLike } from 'fs-extra';
import { ThunkAction } from 'redux-thunk';
import { Repository, Card, Metafile, Error, UUID } from '../types';
import { RootState } from '../store/root';
import { AnyAction } from 'redux';
import { getRepoRoot, isValidRepository, extractRepoName, extractFromURL } from './git';
import { v4 } from 'uuid';
import parsePath from 'parse-path';
import { ActionKeys, NarrowActionType, Action } from '../store/actions';
import { getMetafile } from './metafiles';

type AddRepoAction = NarrowActionType<ActionKeys.ADD_REPO>;
type UpdateRepoAction = NarrowActionType<ActionKeys.UPDATE_REPO>;
type AddErrorAction = NarrowActionType<ActionKeys.ADD_ERROR>;
type UpdateCardAction = NarrowActionType<ActionKeys.UPDATE_CARD>;

/**
 * Action Creator for composing a valid ADD_REPO Redux Action.
 * @param repo A Repository object containing a valid UUID.
 * @return An `AddRepoAction` object that can be dispatched via Redux.
 */
const addRepository = (root: string, { url, oauth }: Partial<ReturnType<typeof extractFromURL>>, username?: string, password?: string):
  AddRepoAction | AddErrorAction => {
  const repo: Repository = {
    id: v4(),
    name: url ? extractRepoName(url.href) : '',
    root: root,
    corsProxy: new URL('https://cors-anywhere.herokuapp.com'), // TODO: This is just a stubbed URL for now, but eventually we need to support Cross-Origin Resource Sharing (CORS) since isomorphic-git requires it
    url: url ? url : parsePath(''),
    local: [],
    remote: [],
    oauth: oauth ? oauth : 'github',
    username: username ? username : '',
    password: password ? password : '',
    token: ''
  };
  if (!isValidRepository(repo)) return reposError(repo.id, `Malformed repository '${repo.name}' cannot be added to the Redux store`);
  return {
    type: ActionKeys.ADD_REPO,
    id: repo.id,
    repo: repo
  };
}

/**
 * Action Creator for composing a valid UPDATE_REPO Redux Action. If the current Redux store does not contain a 
 * matching repository (based on UUID) for the passed parameter, then dispatching this action will not result in any
 * changes in the Redux store state.
 * @param repo A Repository object containing new field values to be updated.
 * @return An `UpdateRepoAction` object that can be dispatched via Redux.
 */
const updateRepository = (repo: Repository): UpdateRepoAction => {
  return {
    type: ActionKeys.UPDATE_REPO,
    id: repo.id,
    repo: repo
  }
}

/**
 * Action Creator for composing a valid ADD_ERROR Redux Action.
 * @param target Corresponds to the object or field originating the error.
 * @param message The error message to be displayed to the user.
 * @return An `AddErrorAction` object that can be dispatched via Redux.
 */
export const reposError = (target: string, message: string): AddErrorAction => {
  const error: Error = {
    id: v4(),
    type: 'ReposError',
    target: target,
    message: message
  };
  return {
    type: ActionKeys.ADD_ERROR,
    id: error.id,
    error: error
  };
}

/**
* Action Creator for composing a valid UPDATE_CARD Redux Action which changes the Metafile associated with the card.
* @param card A `Card` object containing new field values to be updated.
* @param metafile: A `Metafile` object that represents the new metafile information for the card.
* @return An `UpdateCardAction` object that can be dispatched via Redux.
*/
const switchCardMetafile = (card: Card, metafile: Metafile): UpdateCardAction => {
  return {
    type: ActionKeys.UPDATE_CARD,
    id: card.id,
    card: {
      ...card,
      name: metafile.name,
      modified: metafile.modified,
      metafile: metafile.id
    }
  }
}

/**
 * Thunk Action Creator for examining and updating the list of Git branch refs associated with a Repository
 * in the Redux store. Any local or remote branches are captured and added to the associated fields in the
 *  Repository, and the Redux store is updated to reflect these changes.
 * @param root The root Git directory path associated with the `Repository` object.
 * @param repo The `Repository` object that needs to be updated with the latest branch `refs` list.
 * @return A Thunk that can be executed to simultaneously dispatch Redux updates and return the updated `Repository` object 
 * from the Redux store.
 */
export const updateBranches = (id: UUID): ThunkAction<Promise<UpdateRepoAction | AddErrorAction>, RootState, undefined, Action> =>
  async (dispatch, getState) => {
    const repo = getState().repos[id];
    if (!repo) return dispatch(reposError(id, `Cannot update non-existing repo for id:'${id}'`));
    const localBranches = await isogit.listBranches({ fs: fs, dir: repo.root.toString() });
    const remoteBranches = await isogit.listBranches({ fs: fs, dir: repo.root.toString(), remote: 'origin' });
    return dispatch(updateRepository({ ...repo, local: localBranches, remote: remoteBranches }));
  };

/**
* Thunk Action Creator for switching Git branches (or commit hash) and updating the associated metafile and card. Multiple cards
* can be associated with a single metafile, therefore switching branches requires a new (or at least different) metafile that
* refers to a new combination of `Repository` and `ref`. Since `isomorphic-git.checkout()` switches all files within a repository,
* we switch to the target branch and read the specific file, before switching back to the original branch.
* @param cardId The UUID associated with the card that will display the updated content.
* @param metafileId The UUID associated with the original metafile.
* @param branch Git branch name or commit hash; defaults to 'master'.
* @param progress Boolean switch to print phase progress information from `isomorphic-git.checkout()` to console.
* @return An Thunk that can be executed to get a Metafile associated with a new Git branch and updates the card.
*/
export const checkoutBranch = (cardId: UUID, metafileId: UUID, branch: string, progress?: boolean):
  ThunkAction<Promise<Metafile | undefined>, RootState, undefined, AnyAction> =>
  async (dispatch, getState) => {
    if (!getState().cards[cardId]) throw new Error('Redux Error: Cannot update a card that does not exist in the store.');
    const metafile = getState().metafiles[metafileId];
    if (!metafile) throw new Error('Redux Error: Cannot update a metafile that does not exist in the store.');

    const repo = metafile.repo ? getState().repos[metafile.repo] : undefined;
    if (!repo) dispatch(repositoryMissingError(metafile));
    if (repo && metafile.path) {
      const baseRef = metafile.branch;
      // change the repository branch to the new branch
      if (progress) await isogit.checkout({ fs: fs, dir: repo.root.toString(), ref: branch, onProgress: (e) => console.log(e.phase) });
      else await isogit.checkout({ fs: fs, dir: repo.root.toString(), ref: branch });

      // get an updated metafile based on the new branch
      const updated = await dispatch(getMetafile({ filepath: metafile.path }));
      if (!updated) {
        dispatch(metafileMissingError(metafile.path.toString()));
        return undefined;
      }
      // update the metafile details (including file content) for the card
      dispatch(switchCardMetafile(getState().cards[cardId], updated));

      // change the repository branch back to the old branch
      await isogit.checkout({ fs: fs, dir: repo.root.toString(), ref: baseRef });
      if (progress) console.log(`checkout complete...`);
    }
    return getState().metafiles[metafile.id];
  };

/**
 * Thunk Action Creator for retrieving a `Repository` object associated with the given filepath. If the filepath is not under version
 * control (i.e. not contained within a Git repository), or the associated Git repository is malformed, then no valid response can be 
 * given. Otherwise, the correct Repository object is returned and the Redux store is updated with the latest set of branches.
 * @param filepath The relative or absolute path to evaluate.
 * @return  A Thunk that can be executed to simultaneously dispatch Redux updates (as needed) and retrieve a `Repository` object, if
 * the filepath is untracked or results in a malformed Git repository then `undefined` will be returned instead.
 */
export const getRepository = (filepath: PathLike): ThunkAction<Promise<Repository | undefined>, RootState, undefined, Action> =>
  async (dispatch, getState) => {
    const root = await getRepoRoot(filepath);
    if (!root) return undefined; // if there is no root, then filepath is not under version control

    const remoteOriginUrls: string[] | undefined = root ?
      await isogit.getConfigAll({ fs: fs, dir: root.toString(), path: 'remote.origin.url' }) : undefined;
    const { url, oauth } = (remoteOriginUrls && remoteOriginUrls?.length > 0) ?
      extractFromURL(remoteOriginUrls[0]) : { url: undefined, oauth: undefined };
    const existing = url ? Object.values(getState().repos).find(r => r.name === extractRepoName(url.href) && r.url.href === url.href) : undefined;
    if (existing) {
      // the associated repository is already available in the Redux store
      await dispatch(updateBranches(existing.id));
      return getState().repos[existing.id];
    }
    const username = root ? await isogit.getConfig({ fs: fs, dir: root.toString(), path: 'user.name' }) : undefined;
    const password = root ? await isogit.getConfig({ fs: fs, dir: root.toString(), path: 'credential.helper' }) : undefined;

    const action = addRepository(root, { url, oauth }, username, password);
    dispatch(action); // dispatches either AddRepoAction or AddError
    if (action.type === ActionKeys.ADD_REPO) {
      await dispatch(updateBranches(action.id));
      return getState().repos[action.id];
    }
    return undefined; // the constructed repository was malformed and not added to the Redux store
  };