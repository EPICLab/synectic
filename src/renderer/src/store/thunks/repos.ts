import type { PathLike } from 'fs-extra';
import type { FilebasedMetafile } from 'types/metafile';
import type { Repository } from 'types/repo';
import type { ExactlyOne } from 'types/util';
import { createAppAsyncThunk } from '../hooks';
import repoSelectors from '../selectors/repos';
import { isVersionedMetafile } from '../slices/metafiles';
import { repoAdded } from '../slices/repos';
import { fetchBranches } from './branches';
import { fetchParentMetafile } from './metafiles';

export const fetchRepo = createAppAsyncThunk<
  Repository | undefined,
  ExactlyOne<{ filepath: PathLike; metafile: FilebasedMetafile }>
>('repos/fetchRepo', async (input, thunkAPI) => {
  const state = thunkAPI.getState();

  if (input.metafile) {
    // if metafile already has a repo UUID, check for matching repository
    let repo = input.metafile.repo
      ? repoSelectors.selectById(state, input.metafile.repo)
      : undefined;
    // otherwise if parent metafile already has a repo UUID, check for matching repository
    const parent = !repo
      ? await thunkAPI.dispatch(fetchParentMetafile(input.metafile)).unwrap()
      : undefined;
    repo =
      parent && isVersionedMetafile(parent) ? repoSelectors.selectById(state, parent.repo) : repo;
    if (repo) return repo;
  }

  // unless filepath has a root path, there is no repository
  const filepath = input.metafile ? input.metafile.path : input.filepath.toString();
  const { dir } = await window.api.git.getWorktreePaths(filepath);
  if (!dir) return undefined;

  // check root for existing repository
  const existingRepo = repoSelectors.selectByRoot(state, dir);
  const repo = existingRepo ? existingRepo : await thunkAPI.dispatch(buildRepo(dir)).unwrap();
  return repo;
});

export const buildRepo = createAppAsyncThunk<Repository, PathLike>(
  'repos/buildRepo',
  async (filepath, thunkAPI) => {
    const { dir } = await window.api.git.getWorktreePaths(filepath.toString());
    const { url, oauth } = await window.api.git.getRemoteConfig(dir);
    if (dir) await window.api.git.worktreePrune({ dir: dir, verbose: true }); // prune worktree information to remove stale linked branches
    const current = dir
      ? await window.api.git.revParse({ dir: dir, opts: ['abbrevRef'], args: ['HEAD'] })
      : undefined;
    const branches = dir
      ? await thunkAPI.dispatch(fetchBranches(dir)).unwrap()
      : { local: [], remote: [] };
    const { local, remote } = {
      local: branches.local.map(branch => branch.id),
      remote: branches.remote.map(branch => branch.id)
    };
    const { username, password } = await window.api.git.getCredentials(dir);

    return thunkAPI.dispatch(
      repoAdded({
        id: window.api.uuid(),
        name: url
          ? window.api.git.extractRepoName(url.href)
          : dir
          ? window.api.fs.extractFilename(dir)
          : '',
        root: dir ?? '',
        corsProxy: 'https://cors-anywhere.herokuapp.com',
        url: url?.href ?? '',
        default: current ?? '',
        local: local,
        remote: remote,
        oauth: oauth ?? 'github',
        username: username,
        password: password,
        token: ''
      })
    ).payload;
  }
);
