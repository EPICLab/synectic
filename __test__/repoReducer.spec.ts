import parsePath from 'parse-path';

import { Repository } from '../src/types';
import { ActionKeys } from '../src/store/actions';
import { reposReducer } from '../src/store/reducers/repos';

describe('repoReducer', () => {
  const repos: { [id: string]: Repository } = {
    '23': {
      id: '23',
      name: 'sampleUser/myRepo',
      root: 'sampleUser/',
      corsProxy: new URL('http://www.oregonstate.edu'),
      url: parsePath('https://github.com/sampleUser/myRepo'),
      refs: ['942043', '234412', '194724'],
      oauth: 'github',
      username: 'sampleUser',
      password: '12345',
      token: '584n29dkj1683a67f302x009q164'
    }
  }

  const newRepo: Repository = {
    id: '17',
    name: 'sampleUser/forkedRepo',
    root: '/',
    corsProxy: new URL('http://www.oregonstate.edu'),
    url: parsePath('https://github.com/sampleUser/forkedRepo'),
    refs: ['601421', '843449'],
    oauth: 'github',
    username: 'sampleUser',
    password: '12345',
    token: 'a78bw2591q592s0996q1498c1284'
  }

  it('repoReducer returns default state when current state is blank', () => {
    const newRepos = reposReducer(undefined, { type: ActionKeys.ADD_REPO, id: newRepo.id, repo: newRepo });
    expect(Object.keys(newRepos)).toHaveLength(1);
    expect(newRepos).toMatchSnapshot();
  });

  it('repoReducer appends a new repo to state on action ADD_REPO', () => {
    const addedRepos = reposReducer(repos, { type: ActionKeys.ADD_REPO, id: newRepo.id, repo: newRepo });
    expect(Object.keys(addedRepos)).toHaveLength(2);
    expect(addedRepos).toMatchSnapshot();
  });

  it('repoReducer removes a repo from state on action REMOVE_REPO', () => {
    const matchedRepos = reposReducer(repos, { type: ActionKeys.REMOVE_REPO, id: '23' });
    expect(Object.keys(matchedRepos)).toHaveLength(0);
  });

  it('repoReducer resolves non-matching repo in state on action REMOVE_REPO', () => {
    const nonMatchedRepos = reposReducer(repos, { type: ActionKeys.REMOVE_REPO, id: newRepo.id });
    expect(Object.keys(nonMatchedRepos)).toHaveLength(Object.keys(repos).length);
  });

  it('repoReducer updates state of matched repo on action UPDATE_REPO', () => {
    const updatedRepos = reposReducer(repos, {
      type: ActionKeys.UPDATE_REPO, id: '23', repo: {
        oauth: 'bitbucket'
      }
    });
    expect(updatedRepos).not.toMatchObject(repos);
    expect(updatedRepos).toMatchSnapshot();
  });
});