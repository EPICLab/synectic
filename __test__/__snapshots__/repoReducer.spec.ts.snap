// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reposReducer reposReducer appends a new repo to state on action ADD_REPO 1`] = `
Object {
  "17": Object {
    "corsProxy": "http://www.oregonstate.edu/",
    "id": "17",
    "local": Array [
      "601421",
      "843449",
    ],
    "name": "sampleUser/forkedRepo",
    "oauth": "github",
    "password": "12345",
    "remote": Array [],
    "root": "/",
    "token": "a78bw2591q592s0996q1498c1284",
    "url": Object {
      "hash": "",
      "href": "https://github.com/sampleUser/forkedRepo",
      "pathname": "/sampleUser/forkedRepo",
      "port": null,
      "protocol": "https",
      "protocols": Array [
        "https",
      ],
      "query": Object {},
      "resource": "github.com",
      "search": "",
      "user": "",
    },
    "username": "sampleUser",
  },
  "23": Object {
    "corsProxy": "http://www.oregonstate.edu/",
    "id": "23",
    "local": Array [
      "942043",
      "234412",
      "194724",
    ],
    "name": "sampleUser/myRepo",
    "oauth": "github",
    "password": "12345",
    "remote": Array [],
    "root": "sampleUser/",
    "token": "584n29dkj1683a67f302x009q164",
    "url": Object {
      "hash": "",
      "href": "https://github.com/sampleUser/myRepo",
      "pathname": "/sampleUser/myRepo",
      "port": null,
      "protocol": "https",
      "protocols": Array [
        "https",
      ],
      "query": Object {},
      "resource": "github.com",
      "search": "",
      "user": "",
    },
    "username": "sampleUser",
  },
}
`;

exports[`reposReducer reposReducer returns default state when current state is blank 1`] = `
Object {
  "17": Object {
    "corsProxy": "http://www.oregonstate.edu/",
    "id": "17",
    "local": Array [
      "601421",
      "843449",
    ],
    "name": "sampleUser/forkedRepo",
    "oauth": "github",
    "password": "12345",
    "remote": Array [],
    "root": "/",
    "token": "a78bw2591q592s0996q1498c1284",
    "url": Object {
      "hash": "",
      "href": "https://github.com/sampleUser/forkedRepo",
      "pathname": "/sampleUser/forkedRepo",
      "port": null,
      "protocol": "https",
      "protocols": Array [
        "https",
      ],
      "query": Object {},
      "resource": "github.com",
      "search": "",
      "user": "",
    },
    "username": "sampleUser",
  },
}
`;

exports[`reposReducer reposReducer updates state of matched repo on action UPDATE_REPO 1`] = `
Object {
  "23": Object {
    "corsProxy": "http://www.oregonstate.edu/",
    "id": "23",
    "local": Array [
      "942043",
      "234412",
      "194724",
    ],
    "name": "sampleUser/myRepo",
    "oauth": "bitbucket",
    "password": "12345",
    "remote": Array [],
    "root": "sampleUser/",
    "token": "584n29dkj1683a67f302x009q164",
    "url": Object {
      "hash": "",
      "href": "https://github.com/sampleUser/myRepo",
      "pathname": "/sampleUser/myRepo",
      "port": null,
      "protocol": "https",
      "protocols": Array [
        "https",
      ],
      "query": Object {},
      "resource": "github.com",
      "search": "",
      "user": "",
    },
    "username": "sampleUser",
  },
}
`;
